domain/resource/state/resource.go

```diff
@@ -407,173 +419,568 @@ WHERE uuid = $resourceIdentity.uuid`,
+	}
+	var t test
+
+	stmt, err := st.Prepare(`
+SELECT &test.charm_modified_version
+FROM   application a
+JOIN   application_resource ar ON ar.application_uuid = a.uuid
+WHERE  resource_uuid = $resourceIdentity.uuid
+`, resID, t)
 	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+		return errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, stmt, resID).Get(&t)
+	if err != nil {
+		return errors.Errorf("updating charm modified version: %w", err)
+	}
+	st.logger.Criticalf("cmv is %d", t.CMV)
+
+	return nil
+}
+
+// SetApplicationResource marks an existing resource as in use by a CAAS
+// application.
+// Returns [resourceerrors.ResourceNotFound] if the resource UUID cannot be
+// found.
+func (st *State) SetApplicationResource(
+	ctx context.Context,
+	resourceUUID coreresource.UUID,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	// Prepare statement to check if the unit/resource is not already there.
+	k8sAppResource := kubernetesApplicationResource{
+		ResourceUUID: resourceUUID.String(),
+		AddedAt:      st.clock.Now(),
+	}
+	checkK8sAppResourceAlreadyExistsStmt, err := st.Prepare(`
+SELECT &kubernetesApplicationResource.*
+FROM   kubernetes_application_resource
+WHERE  kubernetes_application_resource.resource_uuid = $kubernetesApplicationResource.resource_uuid
+`, k8sAppResource)
+	if err != nil {
+		return errors.Capture(err)
 	}
 
 	// Prepare statement to insert a new link between unit and resource.
-	insertUnitResourceQuery := `
-INSERT INTO unit_resource (unit_uuid, resource_uuid, added_at)
-VALUES ($unitResource.*)`
-	insertUnitResourceStmt, err := st.Prepare(insertUnitResourceQuery, unitResourceInput)
+	insertK8sAppResourceStmt, err := st.Prepare(`
+INSERT INTO kubernetes_application_resource (resource_uuid, added_at)
+VALUES      ($kubernetesApplicationResource.*)
+`, k8sAppResource)
 	if err != nil {
-		return resource.SetUnitResourceResult{}, errors.Capture(err)
+		return errors.Capture(err)
 	}
 
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
 		// Check unit resource is not already inserted.
-		err := tx.Query(ctx, checkUnitResourceStmt, unitResourceInput).Get(&unitResourceInput)
+		err := tx.Query(ctx, checkK8sAppResourceAlreadyExistsStmt, k8sAppResource).Get(&k8sAppResource)
 		if err == nil {
-			return nil // nothing to do
+			// If the kubernetes application resource already exists, do nothing
+			// and return.
+			return nil
 		}
 		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
 			return errors.Capture(err)
 		}
 
-		// Check resource and unit exists.
-		err = tx.Query(ctx, checkValidResourceStmt, resourceUUID).Get(&resourceUUID)
-		if errors.Is(err, sqlair.ErrNoRows) {
-			return errors.Errorf("resource %s: %w", resourceUUID.UUID, resourceerrors.ResourceNotFound)
-		}
+		// Check that this resource exists and is a container image resource.
+		// Only container image resources can be set as kubernetes application
```

Comment:
I'm not sure this is strictly true. It may be logically true?  Where have you seen it codified?

Reply:
Ah this might be my mistake. I had assumed this table was only for recording when the OCI image resources are used, similarly to how the unit table records when a unit is using, which I had also assumed could only be a file type resource. Lets sync on this because it would be good to check my understanding here.

Reply:
Logically you may be correct. However I haven't seen any code which explicitly disallows either type of resource on either machines or kubernetes applications. 

Let's not add an unexpected constraint in a change which does not cover it unless we can find proof.

Reply:
They have been removed.